### PR TITLE
feat(stdlib): Optimize string.trim()

### PR DESCRIPTION
**Pull request by spotandjake**
closes: #982

### CODE REVIEW

File: stdlib/string.gr [hunk 1]
```diff
@@ -1610,30 +1610,40 @@ export let rec forEachCodePointi = (fn: (Number, Number) -> Void, str: String) =
...
+    currentChar == '\u{0009}'  || // Tab
+    currentChar == '\u{000B}'  || // LINE TABULATION
+    currentChar == '\u{000C}'  || // FORM FEED (FF)
+    currentChar == '\u{0020}'  || // Space
+    currentChar == '\u{00A0}'  || // No Break Space
+    currentChar == '\u{FEFF}'  || // ZERO WIDTH NO-BREAK SPACE
+    // Line Terminators
+    currentChar == '\n' || // LF
+    currentChar == '\r' // CR
+  }
```
[the remark this answers is not in the document]
Comment (peblair) [reply stdlib/string.gr]: Yes, @cician is correct for the moment (as far as I can recall). We'll ideally want to have proper lambda lifting in the future. See #1013.

File: stdlib/string.gr [hunk 2]
```diff
@@ -1610,30 +1610,40 @@ export let rec forEachCodePointi = (fn: (Number, Number) -> Void, str: String) =
...
-      currentChar != '\u{FEFF}'  && // ZERO WIDTH NO-BREAK SPACE
-      // Line Terminators
-      currentChar != '\n' && // LF
-      currentChar != '\r' // CR
-    ) break
-  }
-  if (end) slice(0, i+1, str)
-  else slice(i, charsLength, str)
+  if (charsLength > 0 && isWhiteSpace(charAt(i, str))) {
+    if (charsLength < 100) {
```
[the remark this answers is not in the document]
Comment (ospencer) [reply stdlib/string.gr:1635]: There's not really a great way to write memory-constrained tests outside of the runtime right now—something we should probably look into though 🙂

**Comment by ospencer** [comment]
Hey @cician and @spotandjake! Sorry I've been busy with travel to chime in here, but I'm back now 🙂 

I think both of you made some great points! @spotandjake you're right that we want to have the code in the standard library overwhelmingly be readable—it's a great place for people to see what Grain code looks like and it should try to set good examples of how we want people to write code. @cician great callouts about how we can make the implementation faster. I've noticed that you're pretty interested in speed—maybe you can help us get some of these compiler optimizations implemented! I mentioned some time ago that we can finally refactor Chars to be stack-allocated rather than heap-allocated... Maybe that's something you could take on!

The general approach we want to take towards low-level optimizations moving forward is having them exist entirely in the  runtime, with stdlib functions using these runtime helpers, or just being re-exported from the runtime. @cician I think it makes sense to implement your ideas in `runtime/string`.

@spotandjake Up to you if you want to rework this implementation a bit before we get a low level version in. Let me know and we can see what we can do 😄

**Comment by spotandjake** [comment]
> Hey @cician and @spotandjake! Sorry I've been busy with travel to chime in here, but I'm back now 🙂
> 
> I think both of you made some great points! @spotandjake you're right that we want to have the code in the standard library overwhelmingly be readable—it's a great place for people to see what Grain code looks like and it should try to set good examples of how we want people to write code. @cician great callouts about how we can make the implementation faster. I've noticed that you're pretty interested in speed—maybe you can help us get some of these compiler optimizations implemented! I mentioned some time ago that we can finally refactor Chars to be stack-allocated rather than heap-allocated... Maybe that's something you could take on!
> 
> The general approach we want to take towards low-level optimizations moving forward is having them exist entirely in the runtime, with stdlib functions using these runtime helpers, or just being re-exported from the runtime. @cician I think it makes sense to implement your ideas in `runtime/string`.
> 
> @spotandjake Up to you if you want to rework this implementation a bit before we get a low level version in. Let me know and we can see what we can do 😄

if you guys want a low level version i should probably let @cician take over he is a lot better at that stuff than me, the only reason i would want to rework is if stack allocated chars get added and that along with a few optimizations to the code based on some of cicians findings mean we are fast enough to not care about the extra bit from a  low level version.

**Comment by ospencer** [comment]
I think we can hold off on trying to optimize it with this approach—let's close for now, and re-evaluate once the char optimization is done. Cician is also welcome to implement the low level version in the runtime to avoid `explode` entirely.

**Comment by cician** [comment]
I'll think a bit about possible low level primitives to build higher level code upon. I'll get back to you on Discord.

**Pull request closed without merging**